### DISCUSSION TITLE
StochasticTools: Trainers/skip_unconverged_samples Issue

### DISCUSSION THREAD

**Discussion by higgnm** [idaholab/moose#18559] [Q&A Modules: General] [answered]
Hello,
I'm having trouble using `skip_unconverged_samples`. I finally got past "must use / delimiter" and "reporter not declared" errors with `converged_reporter`, but I am now getting an error:
> Reporter with the name "stoch_tools_converged/bool" within app "pr_sub093" was not found.

Which seems to stem from my chosen reporter. I'm using a PostProcessor that simply checks if the transient has run for 60+ seconds to determine that it has "converged":

```
[stoch_tools_converged]
    type = PostprocessorComparison
    comparison_type = greater_than_equals
    value_a = time
    value_b = 60
[]
```

If I change `bool` to `value`, I correctly get an error stating that the reporter has been requested of type `std::vector<bool>` but declared of type `std::vector<double>`. This indicates to me that my usage is fundamentally wrong. Based off the wording, I would think `converged_reporter` requires a bool, but I can't find any tests demonstrating usage of these keywords to guide me. Any help would be appreciated.

Also, as a side note, if  StochasticTools can print `multiapp_converged` in output .csv files, why can't `skip_unconverged_samples` simply be set to `true` without need for additional manipulation?

**Comment by GiudGiud** [comment]
@zachmprince

**Comment by zachmprince** [comment, marked as answer]
Hi @higgnm, you seem to have found a very new capability in stochastic tools! Could you post your inputs to see exactly what you are trying to do?

To summarize how `skip_unconverged_samples` works. First, [SamplerReporterTransfer](https://mooseframework.inl.gov/source/transfers/SamplerReporterTransfer.html) automatically creates a reporter value of type `std::vector<bool>` which contains a value for each sample, `true` if the solve was able to converge and `false` if it doesn't. Whether the solve converged or not directly determined from the sub-app's nonlinear system, there is no need for a separate postprocessor or reporter to determine this. Second, these values are transferred to a [StochasticReporter](https://mooseframework.inl.gov/source/reporters/StochasticReporter.html) so the resulting name is `<stochastic_reporter_name>/<transfer_name>:converged`. Finally, the trainer has the option to skip samples where the solve did not converge by setting `skip_unconverged_samples = true` and `converged_reporter = <stochastic_reporter_name>/<transfer_name>:converged`.

**Reply by higgnm** [reply]
That worked to get my cases running. Unfortunately I'm now getting a "Task add_vector_postprocessor is not registered to build Reporter derived objects" error which I believe relates to a new version of my program and is a separate issue. Your explanation of the reporter usage makes sense; I was mistaken on how to access the :converged value.

**Reply by GiudGiud** [reply]
For which object are you getting this?
You need to move this object from the [VectorPostprocessor] block to the [Postprocessor] block I think

**Reply by higgnm** [reply]
This error was arising from using a newer version (with the `skip_unconverged_samples`) without updating the surrogate evaluation syntax to the newer Reporter syntax.
